Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLPreprocess.py

```diff
@@ -397,67 +350,111 @@ def _foregroundWidths(self):
+        linePosition = 0.0
+        if not self.getProperty(Prop.LINE_POSITION_INPUT).isDefault:
+            linePosition = self.getProperty(Prop.LINE_POSITION_INPUT).value
         else:
-            if not self.getProperty(Prop.BEAM_CENTRE).isDefault:
-                peakPos = self.getProperty(Prop.BEAM_CENTRE).value
-                beamPosWS = self._createFakePeakPositionTable(peakPos)
-            else:
-                # Beam position will be fitted later.
-                beamPosWS = None
-        return ws, beamPosWS
+            # Fit peak position
+            peakWSName = self._names.withSuffix('peak')
+            xmin = self.getProperty(Prop.XMIN).value
+            if not xmin == Property.EMPTY_DBL:
+                xmin = self.getProperty(Prop.XMIN).value
+                xmin = common.inTOF(xmin, l1, l2, theta)
+            xmax = self.getProperty(Prop.XMAX).value
+            if not self.getProperty(Prop.XMAX).isDefault:
+                xmax = self.getProperty(Prop.XMAX).value
+                xmax = common.inTOF(xmax, l1, l2, theta)
+            peak = FindReflectometryLines(
+                InputWorkspace=ws,
+                OutputWorkspace=peakWSName,
+                RangeLower=xmin,
+                RangeUpper=xmax,
+                StartWorkspaceIndex=self.getProperty(Prop.START_WS_INDEX).value,
+                EndWorkspaceIndex=self.getProperty(Prop.END_WS_INDEX).value
+            )
+            linePosition = peak.LineCentre
+            self._cleanup.cleanup(peak.OutputWorkspace)
+        return ws, linePosition
+
+    def _addSampleLogInfo(self, ws, linePosition):
+        """Add foreground indices and linePosition to the sample logs, names start with reduction."""
+        # Add the fractional workspace index of the beam position to the sample logs of ws.
+        run = ws.run()
+        run.addProperty(common.SampleLogs.LINE_POSITION, float(linePosition), True)
+        # Add foreground start and end workspace indices to the sample logs of ws.
+        hws = self._foregroundWidths()
+        beamPosIndex = int(numpy.rint(linePosition))
+        if beamPosIndex > 255:
+            beamPosIndex = 255
+            self.self.log().warning('Is it a monitor spectrum?')
+        sign = self._workspaceIndexDirection(ws)
+        startIndex = beamPosIndex - sign * hws[0]
+        endIndex = beamPosIndex + sign * hws[1]
+        if startIndex > endIndex:
+            endIndex, startIndex = startIndex, endIndex
+        run.addProperty(common.SampleLogs.FOREGROUND_START, int(startIndex), True)
+        run.addProperty(common.SampleLogs.FOREGROUND_CENTRE, int(beamPosIndex), True)
+        run.addProperty(common.SampleLogs.FOREGROUND_END, int(endIndex), True)
+
+    def _moveDetector(self, ws, linePosition):
+        """Perform detector position correction for direct and reflected beams."""
+        if self.getProperty(Prop.DIRECT_LINE_WORKSPACE).isDefault:
+            return ws
```

Comment:
I don't think we can just return here. We haven't had the change to use the fitted line position yet!

Reply:
Initially, I expected the input workspace be ready (all sample logs exist and detector moved). But it is a good choice to allow freedom here and neither expect the sample log line position nor the detector to be moved. Documentation is missing here. As well as the correct name for 2theta in the documentation as I see.